crates/wasm-smith/src/core/code_builder.rs

```diff
@@ -1080,23 +1124,128 @@ impl CodeBuilder<'_> {
-            instructions.push(Instruction::Drop);
+        let operands = operands.to_vec();
+        let mut operands = operands.as_slice();
+        let label_results = label.results.to_vec();
+        while operands.len() > label_results.len() {
+            self.drop_operand(u, *operands.last().unwrap(), instructions);
             operands = &operands[..operands.len() - 1];
         }
-        for (i, expected) in label.results.iter().enumerate() {
+        for (i, expected) in label_results.iter().enumerate() {
             if let Some(actual) = operands.get(i) {
                 if Some(*expected) == *actual {
                     continue;
                 }
-                for _ in operands[i..].iter() {
-                    instructions.push(Instruction::Drop);
+                for ty in operands[i..].iter().rev() {
+                    self.drop_operand(u, *ty, instructions);
                 }
                 operands = &[];
             }
             instructions.push(arbitrary_val(*expected, u));
         }
     }
+
+    fn drop_operand(
+        &mut self,
+        u: &mut Unstructured<'_>,
+        ty: Option<ValType>,
+        instructions: &mut Vec<Instruction>,
+    ) {
+        if !self.mix_operand_into_global(u, ty, instructions) {
+            instructions.push(Instruction::Drop);
+        }
+    }
+
+    /// Attempts to drop the top operand on the stack by "mixing" it into a
+    /// global.
+    ///
+    /// This is done to avoid dropping values on the floor to ensure that
+    /// everything is part of some computation somewhere. Otherwise, for
+    /// example, most function results are dropped on the floor as the stack
+    /// won't happen to match the function type that we're generating.
+    ///
+    /// This will return `true` if the operand has been dropped, and `false` if
+    /// it didn't for one reason or another.
+    fn mix_operand_into_global(
+        &mut self,
+        u: &mut Unstructured<'_>,
+        ty: Option<ValType>,
+        instructions: &mut Vec<Instruction>,
+    ) -> bool {
+        // If the type of this operand isn't known, for example if it's relevant
+        // to unreachable code, then it can't be combined, so return `false`.
+        let ty = match ty {
+            Some(ty) => ty,
+            None => return false,
+        };
+
+        // Use the input stream to allow a small chance of dropping the value
+        // without combining it.
+        if u.ratio(1, 100).unwrap_or(false) {
+            return false;
+        }
+
+        // Depending on the type lookup or inject a global to place this value
+        // into.
+        let (global, combine) = match ty {
+            ValType::I32 => {
+                let global = *self.allocs.global_dropped_i32.get_or_insert_with(|| {
+                    self.allocs.new_globals.push((ty, ConstExpr::i32_const(0)));
+                    inc(&mut self.allocs.globals_cnt)
+                });
+                (global, Instruction::I32Add)
```

Comment:
I have an intuition that `xor` might be better than `add` for these, because with `add` we start losing bits of information when they overflow and disappear.